### DISCUSSION TITLE
Abstractions Command, Event

### DISCUSSION THREAD

**Discussion by clojj** [fraktalio/fmodel#36] [General]
Hi, great work.. not fully digested fmodel yet, but it is a valuable effort for sure!

I wonder why sealed base-types Event and Command are not in the library.. any specific reason for this ?

Best, Jörg

**Comment by idugalic** [comment]
Hi @clojj,

Sealed/Abstract base-types Event and Command are abstract, but these are domain/case-specific. The library will stay more general and give users the freedom to model the message (commands/C, events/E) as required by a specific domain/business.

I hope it makes sense?

Modeling OR (SUM) relationship for Commands and Events makes sense to me, in most of the scenarios.
The OR relationship could be modeled:

- Sealed/Abstract inheritance (polymorphism)
- Enums
- Union types (if kotlin adds them in the future )

This is a second reason why Sealed/Abstract Command and Event are not in the lib. There are other ways to model OR (SUM)

**Reply by clojj** [reply]
Although modelling C and E any other way than with sealed types seems questionable to me, I can see your argument.
Recently, sealed _interfaces_ where introduced for allowing libraries to let users implement their sumtypes elsewhere.. so that was one of the reasons I asked.

**Reply by idugalic** [reply]
Understood, I am with you. I would suggest keeping the library more general (for now), allowing users to model the messages (C, E) on their own. Thanks for initiating this discussion @clojj. Best, Ivan
